torch/_subclasses/fake_tensor.py

```diff
@@ -1313,6 +1314,49 @@ def tolist(self):
 # memory should not significantly increase.
 
 
+# Note - [On fake tensor policy and fresh fake modes for backends]
+#
+# FakeTensorMode does memoization - this memoization is generally fine, but there are some cases
+# where we want to avoid this memoization in favor of producing fake tensors anew. In dynamo, the
+# case for when this happens is when we call a backend. All backends are invoked with a fresh fake_mode
+# because after dynamo trace, the memoized tensors reflect the state of the fake tensor *at the end* of
+# trace, rather than at the beginning.
+# Consider a motivating example of .data setting mutating metadata
+#
+# def foo(x, y):
+#     x.data = y
+#     return x
+#
+# Where x is size([6]) and y is size([3]). If we run foo(x, y), then x.data is size([3]), and the
+# fake tensor at the beginning of our backend, as memoized, is size([3]). However, this means that
+# the backend sees a tensor of size([3]) which has been resized by the user, and so is not reflective
+# of the state of the tensor at the start of trace! In the case of aot_autograd, this causes us to produce
+# incorrect code (concretely in this example, a view into x sized at ([3]))
+#
+#
+# If we do not faithfully preserve those policy decisions through to the new fake_mode, we will produce
+# fake tensors in a slightly different way, with different dynamic indices, which in turn may create new symbols
+# where there should not be any.
+#
+# The solution, therefore, is a combination of a fresh fake mode for backends, with a shared shape_env
```

Comment:
What would break if we instead maintained a set of Tensors which have had metadata mutated, and if you try to convert a tensor which has a corresponding metadata tensor that has been mutated return a fresh tensor instead ? IMO this would be simpler